src/utils/useWaypoints.ts

```
@@ -0,0 +1,83 @@
+import { Waypoint } from "../types/Waypoint";
+import {
+    createWaypointQuery,
+    deleteWaypointQuery,
+    getWaypointsQuery,
+    reorderWaypointsQuery,
+    updateWaypointQuery,
+} from "../api/endpoints";
+import { useAppDispatch, useAppSelector } from "../store/store";
+import { selectWaypoints, setWaypoints as setWaypointsSlice } from "../store/slices/appSlice";
+
+export const useWaypoints = () => {
+    // Null signifies that waypoints have not been fetched yet
+    const dispatch = useAppDispatch();
+    const waypoints = useAppSelector(selectWaypoints);
+    const setWaypoints = (waypoints: Waypoint[]) => dispatch(setWaypointsSlice(waypoints));
+
+    const fetchAndSetWaypoints = async () => setWaypoints(await getWaypointsQuery());
+    useEffect(() => {
+        fetchAndSetWaypoints();
+    }, []);
+
+    const createWaypoint = async (waypoint: Waypoint) => {
+        if (waypoints !== null) {
+            waypoint.order = waypoint.order || waypoints.length;
+
+            const response = await createWaypointQuery(waypoint);
+            // Set the waypoint's id to the one generated by the server
+            waypoint.id = response.data.id;
+
+            setWaypoints([...waypoints, waypoint]);
+        }
+    };
+
+    const deleteWaypoint = async (id: string) => {
+        await deleteWaypointQuery(id);
+        setWaypoints(waypoints?.filter((waypoint) => waypoint.id !== id) || []);
+    };
+
+    const editWaypoint = async (waypoint: Waypoint) => {
+        // Update needs to be instant so that changes update on the map when we drag
+        // Before the server responds
+        // So if stuff goes wrong, we roll the changes back
+        const waypointSnapshot = [...waypoints];
+        try {
+            setWaypoints(waypoints?.map((w) => (w.id === waypoint.id ? waypoint : w)) || []);
+            await updateWaypointQuery(waypoint);
+        } catch {
+            setWaypoints(waypointSnapshot);
+        }
+    };
```

In the future can consider updating this behaviour to show a error popup if it fails instead of silently failing (with the only way to tell being that the changes are reverted).